Server crash with too many database connections

## What went wrong? 😲

With too many users accessing an oms instance at once it is possible to run out of database connections and crash the server.

<details>

<summary>Stack Trace</summary>

```
2025-12-08 17:06:51.121162000 [ERROR] <log_panics:130>:thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: DBError { msg: "Failed to open Connection", extra: "Error(None)" }': service/src/sync/mod.rs:150
   0: backtrace::backtrace::libunwind::trace
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.74/src/backtrace/libunwind.rs:116:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.74/src/backtrace/mod.rs:66:5
   1: backtrace::backtrace::trace
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.74/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.74/src/capture.rs:292:9
   3: backtrace::capture::Backtrace::new
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.74/src/capture.rs:257:22
   4: <backtrace::capture::Backtrace as core::default::Default>::default
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.74/src/capture.rs:517:9
   5: log_panics::Config::install_panic_hook::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/log-panics-2.1.0/src/lib.rs:115:29
   6: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/alloc/src/boxed.rs:1980:9
      std::panicking::rust_panic_with_hook
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:841:13
   7: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:706:13
   8: std::sys::backtrace::__rust_end_short_backtrace
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/sys/backtrace.rs:168:18
   9: __rustc::rust_begin_unwind
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:697:5
  10: core::panicking::panic_fmt
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/panicking.rs:75:14
  11: core::result::unwrap_failed
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/result.rs:1732:5
  12: core::result::Result<T,E>::unwrap
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/result.rs:1137:23
      service::sync::is_initialised
             at /Users/brian/Documents/GitHub/open-msupply.worktrees/origin-Bench-server/server/service/src/sync/mod.rs:150:15
  13: service::sync::file_sync_driver::FileSyncDriver::run::{{closure}}
             at /Users/brian/Documents/GitHub/open-msupply.worktrees/origin-Bench-server/server/service/src/sync/file_sync_driver.rs:76:16
  14: server::start_server::{{closure}}::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/macros/select.rs:557:49
  15: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/future/poll_fn.rs:58:9
  16: server::start_server::{{closure}}
             at /Users/brian/Documents/GitHub/open-msupply.worktrees/origin-Bench-server/server/server/src/lib.rs:339:5
  17: remote_server::main::{{closure}}
             at /Users/brian/Documents/GitHub/open-msupply.worktrees/origin-Bench-server/server/server/src/main.rs:14:40
  18: <tokio::task::local::RunUntil<T> as core::future::future::Future>::poll::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/task/local.rs:995:42
  19: tokio::task::local::LocalSet::with::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/task/local.rs:747:13
  20: std::thread::local::LocalKey<T>::try_with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:315:12
  21: std::thread::local::LocalKey<T>::with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:279:15
  22: tokio::task::local::LocalSet::with
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/task/local.rs:745:9
  23: <tokio::task::local::RunUntil<T> as core::future::future::Future>::poll
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/task/local.rs:985:9
  24: tokio::task::local::LocalSet::run_until::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/task/local.rs:639:19
  25: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/future.rs:124:9
  26: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/future.rs:124:9
  27: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:696:57
  28: tokio::runtime::coop::with_budget
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/coop.rs:107:5
      tokio::runtime::coop::budget
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/coop.rs:73:5
      tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:696:25
  29: tokio::runtime::scheduler::current_thread::Context::enter
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:423:19
  30: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:695:36
  31: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:774:68
  32: tokio::runtime::context::scoped::Scoped<T>::set
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/context/scoped.rs:40:9
  33: tokio::runtime::context::set_scheduler::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/context.rs:180:26
  34: std::thread::local::LocalKey<T>::try_with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:315:12
  35: std::thread::local::LocalKey<T>::with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:279:15
  36: tokio::runtime::context::set_scheduler
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/context.rs:180:9
  37: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:774:27
  38: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:683:19
  39: tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:191:28
  40: tokio::runtime::context::runtime::enter_runtime
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/context/runtime.rs:65:16
  41: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/scheduler/current_thread/mod.rs:179:9
  42: tokio::runtime::runtime::Runtime::block_on_inner
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/runtime.rs:361:47
  43: tokio::runtime::runtime::Runtime::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/runtime/runtime.rs:333:13
  44: tokio::task::local::LocalSet::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.40.0/src/task/local.rs:596:9
  45: actix_rt::runtime::Runtime::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/actix-rt-2.10.0/src/runtime.rs:138:9
  46: actix_rt::system::SystemRunner::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/actix-rt-2.10.0/src/system.rs:244:9
  47: remote_server::main
             at /Users/brian/Documents/GitHub/open-msupply.worktrees/origin-Bench-server/server/server/src/main.rs:5:1
  48: core::ops::function::FnOnce::call_once
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  49: std::sys::backtrace::__rust_begin_short_backtrace
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/backtrace.rs:152:18
  50: std::rt::lang_start::{{closure}}
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/rt.rs:199:18
  51: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/ops/function.rs:284:13
      std::panicking::try::do_call
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:589:40
      std::panicking::try
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:552:19
      std::panic::catch_unwind
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panic.rs:359:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/rt.rs:168:24
      std::panicking::try::do_call
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:589:40
      std::panicking::try
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:552:19
      std::panic::catch_unwind
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panic.rs:359:14
      std::rt::lang_start_internal
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/rt.rs:164:5
  52: std::rt::lang_start
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/rt.rs:198:5
  53: _main
```

</details>

## Expected behaviour 🤔

For the server to not crash.

## How to Reproduce 🔨

To reproduce I used the [Bench-server](/msupply-foundation/open-msupply/tree/Bench-server) branch and ran it concurrently in 100 processes (`seq 100 | parallel -j 100 cargo run`). Most of the time the server returns an error instead of crashing however with enough clients the server crashes.

## Your environment 🌱

- Open mSupply Version: `Bench-server` branch
- Platform: macOS
- Database type: SQLite

I managed to reproduce this with Postgres as well. It's sometimes quite hard to crash the server with Postgres - even with 100 benching processes. Other times it crashes with only 50, seems to be quite random. To make it easier to reproduce the max database connections can be set to 1-3. With a max of one connection the server always crashes on startup after the connection timout. With 2-3 it can crash with only one benching process.

<details>

<summary>Crash logs for Postgres</summary>

```
2025-12-09 15:57:35.140138000 [ERROR] <log_panics:130>:thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: DBError { msg: "Failed to open Connection", extra: "Error(None)" }': service/src/sync/synchroniser_driver.rs:123
   0: backtrace::backtrace::libunwind::trace
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.76/src/backtrace/libunwind.rs:117:9
      backtrace::backtrace::trace_unsynchronized
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.76/src/backtrace/mod.rs:66:14
   1: backtrace::backtrace::trace
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.76/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.76/src/capture.rs:294:9
   3: backtrace::capture::Backtrace::new
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.76/src/capture.rs:259:22
   4: <backtrace::capture::Backtrace as core::default::Default>::default
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/backtrace-0.3.76/src/capture.rs:523:9
   5: log_panics::Config::install_panic_hook::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/log-panics-2.1.0/src/lib.rs:115:29
   6: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/alloc/src/boxed.rs:1980:9
      std::panicking::rust_panic_with_hook
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:841:13
   7: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:706:13
   8: std::sys::backtrace::__rust_end_short_backtrace
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/sys/backtrace.rs:168:18
   9: __rustc::rust_begin_unwind
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:697:5
  10: core::panicking::panic_fmt
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/panicking.rs:75:14
  11: core::result::unwrap_failed
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/result.rs:1732:5
  12: core::result::Result<T,E>::unwrap
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/result.rs:1137:23
      service::sync::synchroniser_driver::get_sync_settings
             at service/src/sync/synchroniser_driver.rs:123:15
  13: service::sync::synchroniser_driver::SynchroniserDriver::sync::{{closure}}
             at service/src/sync/synchroniser_driver.rs:98:35
  14: service::sync::synchroniser_driver::SynchroniserDriver::run::{{closure}}
             at service/src/sync/synchroniser_driver.rs:57:55
  15: server::start_server::{{closure}}::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/macros/select.rs:708:49
  16: <core::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/poll_fn.rs:151:9
  17: server::start_server::{{closure}}
             at server/src/lib.rs:387:5
  18: remote_server::main::{{closure}}
             at server/src/main.rs:25:40
  19: <tokio::task::local::RunUntil<T> as core::future::future::Future>::poll::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/local.rs:1049:42
  20: tokio::task::local::LocalSet::with::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/local.rs:795:13
  21: std::thread::local::LocalKey<T>::try_with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:315:12
  22: std::thread::local::LocalKey<T>::with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:279:15
  23: tokio::task::local::LocalSet::with
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/local.rs:793:9
  24: <tokio::task::local::RunUntil<T> as core::future::future::Future>::poll
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/local.rs:1039:9
  25: tokio::task::local::LocalSet::run_until::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/local.rs:691:19
  26: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/future.rs:124:9
  27: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/future.rs:124:9
  28: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:742:54
  29: tokio::task::coop::with_budget
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/coop/mod.rs:167:5
      tokio::task::coop::budget
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/coop/mod.rs:133:5
      tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:742:25
  30: tokio::runtime::scheduler::current_thread::Context::enter
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:432:19
  31: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:741:36
  32: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:829:68
  33: tokio::runtime::context::scoped::Scoped<T>::set
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/context/scoped.rs:40:9
  34: tokio::runtime::context::set_scheduler::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/context.rs:176:26
  35: std::thread::local::LocalKey<T>::try_with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:315:12
  36: std::thread::local::LocalKey<T>::with
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:279:15
  37: tokio::runtime::context::set_scheduler
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/context.rs:176:9
  38: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:829:27
  39: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:729:19
  40: tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:200:28
  41: tokio::runtime::context::runtime::enter_runtime
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/context/runtime.rs:65:16
  42: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/scheduler/current_thread/mod.rs:188:9
  43: tokio::runtime::runtime::Runtime::block_on_inner
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/runtime.rs:356:47
  44: tokio::runtime::runtime::Runtime::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/runtime/runtime.rs:328:13
  45: tokio::task::local::LocalSet::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.47.1/src/task/local.rs:648:9
  46: actix_rt::runtime::Runtime::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/actix-rt-2.11.0/src/runtime.rs:138:9
  47: actix_rt::system::SystemRunner::block_on
             at /Users/brian/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/actix-rt-2.11.0/src/system.rs:240:9
  48: remote_server::main
             at server/src/main.rs:14:1
  49: core::ops::function::FnOnce::call_once
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  50: std::sys::backtrace::__rust_begin_short_backtrace
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/backtrace.rs:152:18
  51: std::rt::lang_start::{{closure}}
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/rt.rs:199:18
  52: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/ops/function.rs:284:13
      std::panicking::try::do_call
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:589:40
      std::panicking::try
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:552:19
      std::panic::catch_unwind
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panic.rs:359:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/rt.rs:168:24
      std::panicking::try::do_call
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:589:40
      std::panicking::try
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:552:19
      std::panic::catch_unwind
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panic.rs:359:14
      std::rt::lang_start_internal
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/rt.rs:164:5
  53: std::rt::lang_start
             at /Users/brian/.rustup/toolchains/1.88-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/rt.rs:198:5
  54: _main
```

</details>

The bencher also sometimes crashes from the server returning a database connection error without crashing the serve:

```
thread 'main' panicked at src/main.rs:437:10:
called `Result::unwrap()` on an `Err` value: TaskError { task_number: 0, description: "token", error: 200 OK

Caused by:
    0: {"data":null,"errors":[{"message":"Failed to open Connection (Error(None))","locations":[{"line":3,"column":3}],"path":["root"]}]}
    1: invalid type: null, expected struct Root at line 1 column 12 }
```

--- a/server/android/src/android.rs
+++ b/server/android/src/android.rs
@@ -56,8 +56,8 @@ pub mod android {
                 host: "n/a".to_string(),
                 database_name: db_path.to_string_lossy().to_string(),
                 database_path: None,
-                connection_pool_max_connections: 10,
-                connection_pool_timeout: 30,
+                connection_pool_max_connections: None,
+                connection_pool_timeout_seconds: None,
                 // See https://github.com/openmsupply/remote-server/issues/1076
                 init_sql: Some(format!("PRAGMA temp_store_directory = '{}';", cache_dir)),
             },

--- a/server/configuration/base.yaml
+++ b/server/configuration/base.yaml
@@ -17,7 +17,7 @@ database:
   password: "password"
   database_name: "omsupply-database"
   connection_pool_max_connections: 10
-  connection_pool_timeout: 30
+  connection_pool_timeout_seconds: 30
 logging:
   #   one of: All | Console | File
   mode: File

--- a/server/configuration/example.yaml
+++ b/server/configuration/example.yaml
@@ -31,7 +31,7 @@
 #   password: "password"
 #   database_name: "omsupply-database"
 #   connection_pool_max_connections: 10
-#   connection_pool_timeout: 30
+#   connection_pool_timeout_seconds: 30
 # logging:
 ##   one of: All | Console | File
 #   mode: Console

--- a/server/repository/src/database_settings.rs
+++ b/server/repository/src/database_settings.rs
@@ -13,6 +13,9 @@ const SQLITE_LOCKWAIT_MS: u32 = 30 * 1000;
 #[cfg(all(not(feature = "postgres"), not(feature = "memory")))]
 const SQLITE_WAL_PRAGMA: &str = "PRAGMA journal_mode = WAL; PRAGMA synchronous = NORMAL;";
 
+const DEFUALT_CONNECTION_POOL_MAX_CONNECTIONS: u32 = 10;
+const DEFAULT_CONNECTION_POOL_TIMEOUT_SECONDS: u64 = 30;
+
 #[derive(Deserialize, Serialize, Clone)]
 pub struct DatabaseSettings {
     pub username: String,
@@ -21,8 +24,8 @@ pub struct DatabaseSettings {
     pub host: String,
     pub database_name: String,
     pub database_path: Option<String>,
-    pub connection_pool_max_connections: u32,
-    pub connection_pool_timeout: u64,
+    pub connection_pool_max_connections: Option<u32>,
+    pub connection_pool_timeout_seconds: Option<u64>,
     /// SQL run once at startup. For example, to run pragma statements
     pub init_sql: Option<String>,
 }
@@ -182,8 +185,16 @@ pub fn get_storage_connection_manager(settings: &DatabaseSettings) -> StorageCon
     }
     info!("Connecting to database '{}'", settings.database_name);
     let pool = Pool::builder()
-        .max_size(settings.connection_pool_max_connections)
-        .connection_timeout(Duration::from_secs(settings.connection_pool_timeout))
+        .max_size(
+            settings
+                .connection_pool_max_connections
+                .unwrap_or(DEFUALT_CONNECTION_POOL_MAX_CONNECTIONS),
+        )
+        .connection_timeout(Duration::from_secs(
+            settings
+                .connection_pool_timeout_seconds
+                .unwrap_or(DEFAULT_CONNECTION_POOL_TIMEOUT_SECONDS),
+        ))
         .build(connection_manager)
         .expect("Failed to connect to database");
     StorageConnectionManager::new(pool)
@@ -199,8 +210,16 @@ pub fn get_storage_connection_manager(settings: &DatabaseSettings) -> StorageCon
         .connection_customizer(Box::new(SqliteConnectionOptions {
             busy_timeout_ms: Some(SQLITE_LOCKWAIT_MS),
         }))
-        .max_size(settings.connection_pool_max_connections)
-        .connection_timeout(Duration::from_secs(settings.connection_pool_timeout))
+        .max_size(
+            settings
+                .connection_pool_max_connections
+                .unwrap_or(DEFUALT_CONNECTION_POOL_MAX_CONNECTIONS),
+        )
+        .connection_timeout(Duration::from_secs(
+            settings
+                .connection_pool_timeout_seconds
+                .unwrap_or(DEFAULT_CONNECTION_POOL_TIMEOUT_SECONDS),
+        ))
         .build(connection_manager)
         .expect("Failed to connect to database");
     StorageConnectionManager::new(pool)
@@ -220,8 +239,8 @@ mod database_setting_test {
             database_name: "".to_string(),
             init_sql,
             database_path: None,
-            connection_pool_max_connections: 10,
-            connection_pool_timeout: 30,
+            connection_pool_max_connections: None,
+            connection_pool_timeout_seconds: None,
         }
     }
 

--- a/server/repository/src/test_db/postgres.rs
+++ b/server/repository/src/test_db/postgres.rs
@@ -24,8 +24,8 @@ pub fn get_test_db_settings(db_name: &str) -> DatabaseSettings {
         database_name: db_name.to_string(),
         init_sql: None,
         database_path: None,
-        connection_pool_max_connections: 10,
-        connection_pool_timeout: 30,
+        connection_pool_max_connections: None,
+        connection_pool_timeout_seconds: None,
     }
 }
 

--- a/server/repository/src/test_db/sqlite.rs
+++ b/server/repository/src/test_db/sqlite.rs
@@ -35,8 +35,8 @@ fn get_test_db_settings_etc(db_name: &str, is_template: bool) -> DatabaseSetting
         },
         init_sql: None,
         database_path: None,
-        connection_pool_max_connections: 10,
-        connection_pool_timeout: 30,
+        connection_pool_max_connections: None,
+        connection_pool_timeout_seconds: None,
     }
 }
 

